v2/yarpctchannel/outbound.go

```diff
@@ -18,90 +18,86 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package tchannel
+package yarpctchannel
 
 import (
 	"context"
+	"io"
 
 	"github.com/uber/tchannel-go"
-	"go.uber.org/yarpc/api/peer"
-	"go.uber.org/yarpc/api/transport"
-	"go.uber.org/yarpc/internal/introspection"
-	intyarpcerrors "go.uber.org/yarpc/internal/yarpcerrors"
-	peerchooser "go.uber.org/yarpc/peer"
-	"go.uber.org/yarpc/peer/hostport"
-	"go.uber.org/yarpc/pkg/errors"
-	"go.uber.org/yarpc/pkg/lifecycle"
-	"go.uber.org/yarpc/yarpcerrors"
+	yarpc "go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/internal/internaliopool"
+	"go.uber.org/yarpc/v2/internal/internalyarpcerror"
+	"go.uber.org/yarpc/v2/yarpcencoding"
+	"go.uber.org/yarpc/v2/yarpcerror"
+	"go.uber.org/yarpc/v2/yarpcpeer"
 )
 
 var (
-	errDoNotUseContextWithHeaders = yarpcerrors.Newf(yarpcerrors.CodeInvalidArgument, "tchannel.ContextWithHeaders is not compatible with YARPC, use yarpc.CallOption instead")
+	errDoNotUseContextWithHeaders = yarpcerror.Newf(yarpcerror.CodeInvalidArgument, "tchannel.ContextWithHeaders is not compatible with YARPC, use yarpc.CallOption instead")
 
-	_ transport.UnaryOutbound              = (*Outbound)(nil)
-	_ introspection.IntrospectableOutbound = (*Outbound)(nil)
+	_ yarpc.UnaryOutbound = (*Outbound)(nil)
 )
 
 // Outbound sends YARPC requests over TChannel.
-// It may be constructed using the NewOutbound or NewSingleOutbound methods on
-// the TChannel Transport.
 type Outbound struct {
-	transport *Transport
-	chooser   peer.Chooser
-	once      *lifecycle.Once
-}
+	// Chooser is a peer chooser for outbound requests.
+	//
+	// The chooser receives the request metadata and returns peers.
+	// The chooser is backed by a dialer and a peer list.
+	// You can instead use an Addr and Dialer directly, bypassing YARPC peer
+	// selection.
+	Chooser yarpc.Chooser
 
-// NewOutbound builds a new TChannel outbound that selects a peer for each
-// request using the given peer chooser.
-func (t *Transport) NewOutbound(chooser peer.Chooser) *Outbound {
-	return &Outbound{
-		once:      lifecycle.NewOnce(),
-		transport: t,
-		chooser:   chooser,
-	}
-}
+	// Addr is the host:port of the peer that will handle outbound requests.
+	//
+	// Providing an Addr and Dialer is an alternative to using YARPC peer
+	// selection.
+	// The outbound will dial this exact address for each outbound request.
+	Addr string
 
-// NewSingleOutbound builds a new TChannel outbound always using the peer with
-// the given address.
-func (t *Transport) NewSingleOutbound(addr string) *Outbound {
-	chooser := peerchooser.NewSingle(hostport.PeerIdentifier(addr), t)
-	return t.NewOutbound(chooser)
-}
+	// Dialer is a dialer to retain connections to the remote peer.
+	//
+	// Providing an Addr and Dialer is an alternative to using YARPC peer
+	// selection.
```

Comment:
Did we eliminate the single chooser? Seems like "provide Chooser or provider
Addr-Dialer" can be changed to "provide Chooser, one of which can be
SinglePeer(dialer, addr)".

I'm okay with keeping this as-is if we want to err on the side of keeping the
user-facing API more easy to use.

Reply:
Yes, I eliminated SingleChooser. This makes for the better developer ergonomics. A couple of design shifts pushed me in this direction. The single peer chooser no longer needed to have lifecycle so it melted into almost nothing. Also, the semantics about retaining and releasing the peer became muddy, since the peer chooser no longer had a life cycle. Passing the burden of retaining and releasing the peer to the user would be hostile to developer experience.

I expect that we will likely want to add a branch here for the case that a peer gets carried on Request. We probably can’t ignore the case of sending an RPC to an exact peer much longer.